src/server/gc_worker/compaction_filter.rs

```diff
@@ -337,86 +405,75 @@ impl Drop for WriteCompactionFilter {
         self.engine.sync_wal().unwrap();
 
         self.switch_key_metrics();
-        STATS.with(|stats| {
+        if let Some((versions, filtered, deleted)) = STATS.with(|stats| {
             stats.versions.update(|x| x + self.total_versions);
             stats.filtered.update(|x| x + self.total_filtered);
             stats.deleted.update(|x| x + self.total_deleted);
-            if stats.versions.get() >= 1024 * 1024
-                || stats.last_report.get().elapsed() >= Duration::from_secs(60)
-            {
-                stats.report();
+            if stats.need_report() {
+                return Some(stats.prepare_report());
             }
-        })
+            None
+        }) {
+            info!(
+                "Compaction filter reports"; "total" => versions,
+                "filtered" => filtered, "deleted" => deleted,
+            );
+        }
     }
 }
 
 impl CompactionFilter for WriteCompactionFilter {
-    fn filter(
+    fn filter_v2(
         &mut self,
-        _start_level: usize,
+        level: usize,
         key: &[u8],
         value: &[u8],
-        _: &mut Vec<u8>,
-        _: &mut bool,
-    ) -> bool {
-        self.near_seek_distance += 1;
-        let (key_prefix, commit_ts) = split_ts(key);
-        if commit_ts > self.safe_point {
-            return false;
+        value_type: CompactionFilterValueType,
+    ) -> CompactionFilterDecision {
+        if self.is_invalid {
+            // If there are already some errors, do nothing.
+            return CompactionFilterDecision::Keep;
         }
 
-        self.versions += 1;
-        if self.key_prefix != key_prefix {
-            self.key_prefix.clear();
-            self.key_prefix.extend_from_slice(key_prefix);
-            self.remove_older = false;
-            self.switch_key_metrics();
-        }
-
-        let mut filtered = self.remove_older;
-        let write = parse_write(value);
-        if !self.remove_older {
-            match write.write_type {
-                WriteType::Rollback | WriteType::Lock => filtered = true,
-                WriteType::Put => self.remove_older = true,
-                WriteType::Delete => {
-                    self.remove_older = true;
-                    filtered = true;
-                    self.handle_delete_mark(key);
-                }
+        match self.do_filter_v2(level, key, value, value_type) {
```

Comment:
We do not need to name it by `do_filter_v2`, because the version is in rocksdb, we do not need to make  it compatible to filter_v1 in rocksdb.
So just name it `filter`.